src/main/java/org/logstash/filewatch/WatchedFilesCollection.java

```diff
@@ -0,0 +1,229 @@
+
+    @JRubyMethod(name = "empty?") // synchronize { @files.empty? }
+    public synchronized IRubyObject empty_p(ThreadContext context) {
+        return context.runtime.newBoolean(this.files.isEmpty());
+    }
+
+    @JRubyMethod
+    public synchronized IRubyObject each_file(ThreadContext context, Block block) {
+        for (IRubyObject watched_file : this.files.keySet()) {
+            block.yield(context, watched_file);
+        }
+        return context.nil;
+    }
+
+    @JRubyMethod // synchronize { @files.values.to_a }
+    public IRubyObject paths(ThreadContext context) {
+        IRubyObject[] values;
+        synchronized (this) {
+            values = this.files.values().stream().toArray(IRubyObject[]::new);
+        }
+        return context.runtime.newArrayNoCopy(values);
+    }
+
+    // NOTE: needs to return properly ordered files (can not use @files_inverse)
+    @JRubyMethod // synchronize { @files.key_set.to_a }
+    public IRubyObject files(ThreadContext context) {
+        IRubyObject[] keys;
+        synchronized (this) {
+            keys = this.files.keySet().stream().toArray(IRubyObject[]::new);
+        }
+        return context.runtime.newArrayNoCopy(keys);
+    }
+
+
+    @JRubyMethod
+    public IRubyObject update(ThreadContext context, IRubyObject file) {
+        // NOTE: modified_at might change on restat - to cope with that we need to potentially
+        // update the sorted collection, on such changes (when file_sort_by: last_modified) :
+        if (!"modified_at".equals(sortBy)) return context.nil;
+
+        RubyString path = getFilePath(context, file);
+        synchronized (this) {
+            this.files.remove(file); // we need to "re-sort" changed file -> remove and add it back
+            modified_at(context, file, context.tru); // file.modified_at(update: true)
+            RubyString prev_path = this.files.put(file, path);
+            assert prev_path == null;
+        }
+        return context.tru;
+    }
+
+    @JRubyMethod(required = 1, visibility = Visibility.PRIVATE)
+    @Override
+    public IRubyObject initialize_copy(IRubyObject original) {
+        final Ruby runtime = getRuntime();
+        if (!(original instanceof WatchedFilesCollection)) {
+            throw runtime.newTypeError("Expecting an instance of class WatchedFilesCollection");
+        }
+
+        WatchedFilesCollection proto = (WatchedFilesCollection) original;
+
+        this.files = new TreeMap<>(proto.files.comparator());
+        synchronized (proto) {
+            this.files.putAll(proto.files);
+            this.filesInverse = (RubyHash) proto.filesInverse.dup(runtime.getCurrentContext());
+        }
+
+        return this;
+    }
+
+    @Override
+    public IRubyObject inspect() {
+        return getRuntime().newString("#<" + metaClass.getRealClass().getName() + ": size=" + this.files.size() + ">");
+    }
+
+    private static final CachingCallSite modified_at_site = new FunctionalCachingCallSite("modified_at");
```

Comment:
I might need you to explain this optimization to me, since `org.jruby.runtime.callsite.CachingCallSite` isn't particularly well-documented 😩

Reply:
kind of does the same as `obj.modified_at` in Ruby, the first time such call-site is executed the target (obj's `class`) and the looked up (`modified_at`) method are cached at the call-site. the next call checks if `obj.class == cached_class` (+ a check that the class hasn't been modified since) and if all is well uses the cached method object to perform the call against, thus avoiding the method lookup.

at the native (Java) end it's pretty much the only API you get for doing the same. 
most of the time one just `obj.callMethod(context, "length")` since the method lookup (hash-map get operation) cost is negligible. 

this path is really hot and it's gets a bit ugly to keep those parts in Ruby land (have tried and it could have been working but would put more burden on trying to reason about the code).
so I've decided to introduce this simple call-site cache here for the sorting operation.
haven't measured how much it's worth (was "late" to deliver the PR already) - just seemed to make sense since the tree-map will constantly call the comparator on `get` as well as `put` ops.

---

the API has been stable for years, it's not super blessed but gets used in places where the native Java parts "really" need to avoid method lookup (for performance reasons). 
it's long due for a better API - mostly for site caching to be unified with invokedynamic. 
still, even if it got deprecated (due a replacement) it's going to be kept as some ext use it.

Reply:
TIL. Thanks for the details!